### PR TITLE
split: meta start partition split

### PR DESCRIPTION
**Pull request by hycdong**
### Simple partition split process
1. **meta receives client partition split request, and change partition count**
2. replica notices partition count changed during on_config_sync
3. parent partitions create child partitions, child partitions async learn mutations and private logs from parents
4. all child partitions in group finish learn process, parent will stop read and write
5. meta server register child partitions
6. child partition active, and parent recover read and write

More partition split discussion in  issue #69 
This pr solves the first step of partition split, which is bold in process description.

### Main code logic
When meta server receives a partition split request:
- meta will firstly validate parameters, including app status, partition count etc
- if validation passed, meta will update new partition_count on zk
- then update new partition_count on meta server locally
- then create child partitions' partition_configuration structure locally, and we assume child partition init balliot = invalid_ballot(-1), which means the partition is currently invalid

### Other explanations
- add init_partition_count in app_info, it will be used when meta server failed during partition split, I will explain it in another new pull request
- add unit tests for function ''app_partition_split"

### CODE REVIEW

File: src/dist/replication/test/meta_test/unit_test/meta_split_service_test_helper.h [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+{
+    dsn::message_ex *binary_req = dsn::message_ex::create_request(rpc_code);
+    dsn::marshall(binary_req, req);
+    return create_corresponding_receive(binary_req);
+}
+
+// send app partition split request
+inline app_partition_split_response send_request(dsn::task_code rpc_code,
+                                                 app_partition_split_request request,
+                                                 std::shared_ptr<meta_service> meta_svc,
```
Comment (neverchanje) [inline src/dist/replication/test/meta_test/unit_test/meta_split_service_test_helper.h]: 你这个写法错了，rpc holder 目的就是让你不需要这么写代码。你想让 meta_split_service 去调用 app_partition_split，然后等：

```cpp
    app_partition_split_response app_partition_split(unique_ptr<app_partition_split_request> req)
    {
        app_partition_split_rpc rpc(std::move(req), RPC_CM_APP_PARTITION_SPLIT);
        split_srv->app_partition_split(rpc.request(), rpc.response());
        return rpc.response();
    }
```

这样写是最简单的。

File: src/dist/replication/meta_server/meta_split_service.cpp [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        response.app_id = app->app_id;
+        response.partition_count = app->partition_count;
+
+        // new_partition_count != old_partition_count*2
+        if (request.new_partition_count != app->partition_count * 2) {
+            response.err = ERR_INVALID_PARAMETERS;
+            dwarn_f("client({}) sent split request with wrong partition count: app({}), partition "
+                    "count({}),"
+                    "new_partition_count({})",
+                    ((message_ex *)rpc.dsn_request())->header->from_address.to_string(),
```
Comment (neverchanje) [inline src/dist/replication/meta_server/meta_split_service.cpp]: 用 `rpc_holder::remote_address`

File: src/dist/replication/meta_server/meta_service.cpp [hunk 3]
```diff
@@ -883,5 +888,21 @@ void meta_service::ddd_diagnose(ddd_diagnose_rpc rpc)
...
 }
 
+void meta_service::on_app_partition_split(app_partition_split_rpc rpc)
+{
+    RPC_CHECK_STATUS(rpc.dsn_request(), rpc.response());
+
+    tasking::enqueue(LPC_META_STATE_NORMAL,
+                     tracker(),
+                     [this, rpc]() {
+                         if (!_split_svc) {
```
Comment (neverchanje) [inline src/dist/replication/meta_server/meta_service.cpp]: 这个没可能发生把

File: src/dist/replication/test/meta_test/unit_test/meta_split_service_test.cpp [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    std::string apps_root = "/meta_test/apps";
+    std::shared_ptr<server_state> ss = meta_svc->_state;
+    ss->initialize(meta_svc.get(), apps_root);
+    ss->_all_apps.emplace(std::make_pair(app->app_id, app));
+    ss->sync_apps_to_remote_storage();
+
+    return meta_svc;
+}
+
+void meta_service_test_app::app_partition_split_test()
```
Comment (neverchanje) [inline src/dist/replication/test/meta_test/unit_test/meta_split_service_test.cpp]: 这种测试的写法是不好的，不同的 test case 不应该通过 cout 分开，你可以参考 meta_duplication_service_test 的写法，应该用 gtest TEST_ 来做

File: src/dist/replication/meta_server/meta_split_service.h [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+namespace dsn {
+namespace replication {
+
+class meta_split_service
+{
+public:
+    explicit meta_split_service(meta_service *meta);
+    // client -> meta to start split
+    void app_partition_split(app_partition_split_rpc rpc);
+    void do_app_partition_split(std::shared_ptr<app_state> app, app_partition_split_rpc rpc);
```
Comment (neverchanje) [inline src/dist/replication/meta_server/meta_split_service.h]: 应该设成 private。

File: src/dist/replication/test/meta_test/unit_test/meta_split_service_test.cpp [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+#include <dsn/service_api_c.h>
+
+#include "dist/replication/meta_server/meta_service.h"
+#include "meta_service_test_app.h"
+#include "meta_split_service_test_helper.h"
+
+using namespace ::dsn::replication;
+
+// create mock meta service
+std::shared_ptr<meta_service> meta_service_test_app::create_mock_meta_svc()
```
Comment (neverchanje) [inline src/dist/replication/test/meta_test/unit_test/meta_split_service_test.cpp]: 你可以把 meta_duplication_service_test 里面重复的代码抽出来，你这个测试写法，相当于测试结束时，meta zk 里面的数据没有清掉，需要 clear.sh 来清理，比较麻烦。。

File: src/dist/replication/replication.thrift [hunk 7]
```diff
@@ -757,6 +757,22 @@ struct ddd_diagnose_response
...
+{
+    1:string                 app_name;
+    2:i32                    new_partition_count;
+}
+
+struct app_partition_split_response
+{
+    1:dsn.error_code         err;
+    2:i32                    app_id;
+    3:i32                    partition_count;
```
[the remark this answers is not in the document]
Comment (hycdong) [reply src/dist/replication/replication.thrift]: It is NOT old partition count, it is just app's partition count. 
If split failed, response.partition_count = original_partition_count,
If split succeed, response.partition_count = new_partition_count ( It is updated in callback of updating partition_count on remote stroage, see line121 in function 'do_app_partition_split')

[the remark this answers is not in the document]
Comment (neverchanje) [reply src/dist/replication/replication.thrift]: OK, you should add these words as comments above `app_partition_count_request`.

File: src/dist/replication/meta_server/meta_split_service.cpp [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+    std::shared_ptr<app_state> app;
+    {
+        zauto_write_lock l(app_lock());
+
+        app = _state->get_app(request.app_name);
+        if (!app || app->status != app_status::AS_AVAILABLE) {
+            response.err = ERR_APP_NOT_EXIST;
+            dwarn_f("client({}) sent split request with invalid app({}), app is not existed or "
+                    "unavailable",
+                    ((message_ex *)rpc.dsn_request())->header->from_address.to_string(),
```
Comment (neverchanje) [inline src/dist/replication/meta_server/meta_split_service.cpp]: rpc.remote_address()

File: src/dist/replication/meta_server/meta_split_service.cpp [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+                                                app_partition_split_rpc rpc)
+{
+    auto on_write_storage_complete = [app, rpc, this]() {
+        ddebug_f("app({}) update partition count on remote storage, new partition count is {}",
+                 app->app_name.c_str(),
+                 app->partition_count * 2);
+
+        zauto_write_lock l(app_lock());
+        app->partition_count *= 2;
+        app->helpers->contexts.resize(app->partition_count);
```
Comment (neverchanje) [inline src/dist/replication/meta_server/meta_split_service.cpp]: 这个 contexts 是干嘛的

Comment (hycdong) [reply src/dist/replication/meta_server/meta_split_service.cpp]: `config_context` struct for each partition, including partition_configuration and other context information, such as if current partition is updating config. More information about this struct can reference [config_context definition](https://github.com/XiaoMi/rdsn/blob/16e633245281ca21c4d46e8852efe27aa46427da/src/dist/replication/meta_server/meta_data.h#L187)

**Pull request merged**